server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3286,19 +3321,101 @@ private void doCheckIndex() throws IOException {
     }
 
     Engine getEngine() {
-        Engine engine = getEngineOrNull();
-        if (engine == null) {
-            throw new AlreadyClosedException("engine is closed");
+        engineLock.readLock().lock();
+        try {
+            return getCurrentEngine(false);
+        } finally {
+            engineLock.readLock().unlock();
         }
-        return engine;
     }
 
     /**
      * NOTE: returns null if engine is not yet started (e.g. recovery phase 1, copying over index files, is still running), or if engine is
      * closed.
      */
     public Engine getEngineOrNull() {
-        return this.currentEngineReference.get();
+        engineLock.readLock().lock();
+        try {
+            return getCurrentEngine(true);
+        } finally {
+            engineLock.readLock().unlock();
+        }
+    }
+
+    public <R> R withMutableEngine(Function<Engine, R> operation) {
+        return withEngine(operation, true, false);
+    }
+
+    public <R> R withMutableEngineOrNull(Function<Engine, R> operation) {
```

Comment:
I think this should better be called `withEngine`. And rename current `withEngine` to something like `withEngineInternal` or something else.